File: leetcode/Python/construct-target-array-with-multiple-sums.py
class Solution:
    def isPossible(self, target: List[int]) -> bool:
        if len(target) == 1:
            return target[0]==1
        curSum, prevSum, num, n = sum(target), 0, 0, 0
        max_heap = [-i for i in target]
        heapq.heapify(max_heap)
        while max_heap[0]<-1:
            prevSum=-heapq.heappop(max_heap)
            n=(prevSum+max_heap[0])//(curSum-prevSum)
            if n<=0:
                n=1
            num=prevSum-n*(curSum-prevSum)
            if num<1:
                return False
            heapq.heappush(max_heap,-num)
            curSum=curSum-prevSum+num
        return True

File: leetcode/C++/construct-target-array-with-multiple-sums.cpp
class Solution {
public:
    bool isPossible(vector<int>& target) {
        if(target.size()==1) return target[0]==1;
        long long curSum=0,prevSum=0,prevNum=0,n;
        priority_queue<int> pq(target.begin(),target.end());
        for(int &t: target) curSum+=t;
        while(!pq.empty()){
            prevSum=pq.top(); pq.pop();
            if(prevSum<=1) break;
            // subtract count until nextPrevSum
            n=(prevSum-pq.top())/(curSum-prevSum);
            // minimum n=1
            if(n<=0) n=1;
            // find previous number
            prevNum=prevSum-(curSum-prevSum)*n;
            // check if previous number is less than 1
            if(prevNum<1) return false;
            pq.push(prevNum);
            // update current sum
            curSum=curSum-prevSum+prevNum;
        }
        return prevSum==1;
    }
};